libsolidity/analysis/TypeChecker.cpp

```diff
@@ -249,7 +249,23 @@ TypePointers TypeChecker::typeCheckABIDecodeAndRetrieveReturnType(FunctionCall c
 	for (auto const& typeArgument: tupleExpression->components())
 	{
 		solAssert(typeArgument, "");
-		if (TypeType const* argTypeType = dynamic_cast<TypeType const*>(type(*typeArgument)))
+		if (
+			!dynamic_cast<ElementaryTypeNameExpression const*>(typeArgument.get()) &&
+			!dynamic_cast<Identifier const*>(typeArgument.get()) &&
+			!dynamic_cast<IndexAccess const*>(typeArgument.get()) &&
+			!dynamic_cast<MemberAccess const*>(typeArgument.get())
+		)
+		{
+
+			m_errorReporter.typeError(
+				5901_error,
+				typeArgument->location(),
+				"Expression is not allowed as argument here. Argument must be "
+				"a built-in type name, a user-defined type name or an array of those."
```

Comment:
```suggestion
				"a built-in type name, a user-defined type name or an array type."
```

When I read "array of those", I think about something like this: `[uint, S, T]`.